File: cmd/recent.ts
// `f5kb recent` — fetch articles modified in the last N days, one JSON per type,
// plus an _index.json manifest. Behavior reference: fetch_recent_by_type.ts.
// The per-type wrapped JSON {documentType, days, cutoff, generatedAt, count,
// articles[]} and _index.json are byte-identical (client-side modMs filter).
//
// Flags:
//   --days=N         REQUIRED window size
//   --out=DIR        REQUIRED output directory
//   --types="A,B"    subset of document types
//   --page-size=N    results per call (default 500, max 1000)
//   --limit=N        cap articles per type (testing)

import { type ParsedArgs } from "../lib/args.ts";
import { applyTypeFilters, flagList, flagNum, flagStr, warnUnknownTypes } from "../lib/args.ts";
import { type Logger } from "../lib/logger.ts";
import { CoveoClient } from "../lib/coveo/client.ts";
import { fetchCoveoConfig, refreshConfig } from "../lib/coveo/aura.ts";
import { fetchFlatChunked, type FlatArticle } from "../lib/coveo/flat.ts";
import { sanitizeName } from "../lib/fsutil.ts";

// fetch_recent_by_type.ts paged with a 120ms inter-page pause (vs flex's 150ms).
const RECENT_PAUSE_MS = 120;

async function listDocumentTypes(
  client: CoveoClient,
): Promise<Array<{ value: string; count: number }>> {
  const values = await client.listFacetValues("f5_document_type");
  return values.filter((v) => !v.value.includes("|"));
}

async function fetchTypeSince(
  client: CoveoClient,
  type: string,
  cutoffMs: number,
  endMs: number,
  pageSize: number,
  limit: number,
): Promise<FlatArticle[]> {
  const baseAq = `@f5_document_type=="${type}"`;
  const collected: FlatArticle[] = [];
  await fetchFlatChunked(
    client,
    baseAq,
    cutoffMs,
    endMs,
    pageSize,
    limit,
    () => {},
    collected,
    RECENT_PAUSE_MS,
  );
  // @date is a superset of the content-mod window; refine to the exact window.
  return collected.filter((a) => a.modMs === undefined || a.modMs >= cutoffMs);
}

export async function run(args: ParsedArgs, logger: Logger): Promise<number> {
  const flags = args.flags;

  const daysRaw = flagStr(flags, "days");
  const days = Number(daysRaw);
  if (!daysRaw || !Number.isFinite(days) || days <= 0) {
    logger.error("--days must be a positive number");
    return 1;
  }
  const outDir = flagStr(flags, "out");
  if (!outDir) {
    logger.error("--out (output directory) is required");
    return 1;
  }
  const pageSize = Math.min(flagNum(flags, "page-size", 500)!, 1000);
  const limit = flags.limit ? parseInt(String(flags.limit)) : Infinity;
  const includeTypes = flagList(flags, "types");
  const excludeTypes = flagList(flags, "exclude-types");

  const nowMs = Date.now();
  const cutoffMs = nowMs - days * 86400000;
  const endMs = nowMs + 86400000;

  logger.info("Fetching Coveo configuration from F5 portal...");
  const coveoConfig = await fetchCoveoConfig();
  logger.info(`Organization ID: ${coveoConfig.organizationId}`);
  const client = new CoveoClient(coveoConfig, {
    logger: logger.child("coveo"),
    refresh: (c) => refreshConfig(c),
  });
  logger.info(
    `Window: articles modified since ${new Date(cutoffMs).toISOString().slice(0, 10)} ` +
      `(last ${days} day${days === 1 ? "" : "s"})`,
  );

  const allTypes = await listDocumentTypes(client);
  const allTypeNames = allTypes.map((t) => t.value);
  if ((includeTypes || excludeTypes)) {
    warnUnknownTypes(allTypeNames, includeTypes, excludeTypes, (m) => {
      logger.warn(m);
      logger.warn(`Known types: ${allTypeNames.join(", ")}`);
    });
  }
  const selected = applyTypeFilters(allTypeNames, includeTypes, excludeTypes);
  if (!selected.length) {
    logger.error("no types selected (after --types / --exclude-types)");
    return 1;
  }

  await Deno.mkdir(outDir, { recursive: true });

  const summary: Array<{ type: string; count: number; file: string }> = [];

  for (const type of selected) {
    const articles = await fetchTypeSince(client, type, cutoffMs, endMs, pageSize, limit);

    const fileName = `${sanitizeName(type)}.json`;
    const filePath = `${outDir}/${fileName}`;
    const payload = {
      documentType: type,
      days,
      cutoff: new Date(cutoffMs).toISOString(),
      generatedAt: new Date(nowMs).toISOString(),
      count: articles.length,
      articles: articles.map(({ name, link, summary, publicationDate, modificationDate }) => ({
        name,
        link,
        summary,
        publicationDate,
        modificationDate,
      })),
    };
    await Deno.writeTextFile(filePath, JSON.stringify(payload, null, 2));

    summary.push({ type, count: articles.length, file: fileName });
    logger.info(
      `${type}: ${articles.length} article${articles.length === 1 ? "" : "s"} -> ${filePath}`,
    );
  }

  const manifestPath = `${outDir}/_index.json`;
  await Deno.writeTextFile(
    manifestPath,
    JSON.stringify(
      {
        days,
        cutoff: new Date(cutoffMs).toISOString(),
        generatedAt: new Date(nowMs).toISOString(),
        totalArticles: summary.reduce((a, s) => a + s.count, 0),
        types: summary,
      },
      null,
      2,
    ),
  );

  const total = summary.reduce((a, s) => a + s.count, 0);
  logger.info(
    `Done. ${total} article${total === 1 ? "" : "s"} across ${selected.length} type(s) ` +
      `written to ${outDir}/ (manifest: ${manifestPath})`,
  );
  return 0;
}

File: f5kb/cmd/recent.py
"""f5kb recent subcommand."""

from __future__ import annotations

import sys
import time

import click

from f5kb.coveo.aura import fetch_coveo_config, refresh_config
from f5kb.coveo.client import CoveoClient
from f5kb.config.loader import load_config
from f5kb.lib.dump import dump_types


@click.command()
@click.option("--days", required=True, type=float)
@click.option("--out", required=True)
@click.option("--types", default=None)
@click.option("--exclude-types", default=None)
@click.option("--page-size", type=int, default=500, show_default=True)
@click.option("--limit", type=int, default=0)
@click.option("--config", "config_path", default="config.yaml", show_default=True)
@click.pass_context
def recent_cmd(ctx, days, out, types, exclude_types, page_size, limit, config_path):
    """Fetch articles modified in the last N days, one JSON per type."""
    log = ctx.obj["logger"]
    try:
        config = load_config(config_path)
    except Exception as e:
        log.error(f"could not read/parse config {config_path}: {e}")
        sys.exit(1)
    type_configs = {k: {"documentType": v.document_type, "metadata": v.metadata, "content": v.content}
                   for k, v in config.types.items()}
    all_type_keys = list(type_configs.keys())
    include_types = [t.strip() for t in types.split(",")] if types else None
    excl_types = [t.strip() for t in exclude_types.split(",")] if exclude_types else None
    type_keys = [t for t in all_type_keys if
                 (not include_types or t in include_types) and
                 (not excl_types or t not in excl_types)]
    if not type_keys:
        log.error("no type keys selected")
        sys.exit(1)

    log.info("Fetching Coveo configuration from F5 portal...")
    try:
        coveo_config = fetch_coveo_config()
    except Exception as e:
        log.error(f"failed to fetch Coveo config: {e}")
        sys.exit(1)
    client = CoveoClient(coveo_config, logger=log.child("coveo"), refresh=lambda c: refresh_config(c))
    now_ms = int(time.time() * 1000)
    cutoff_ms = int(now_ms - days * 86400000)

    result = dump_types(
        client,
        type_configs=type_configs,
        type_keys=type_keys,
        descriptions=dict(config.field_descriptions),
        out_dir=out,
        all_time=False,
        mode=f"days={days}",
        cutoff_ms=cutoff_ms,
        end_ms=now_ms + 86400000,
        now_ms=now_ms,
        page_size=min(page_size, 500),
        limit=limit,
        config_path=config_path,
        logger=log,
    )
    log.info(f"Done. {result.total} articles -> {out}/")
